### PR TITLE
PMTiles + enhancements to vector tile data sources in Rapid

### PR DESCRIPTION
**Pull request by bhousel**
### tl;dr

This PR includes a massive refresh of the vector tile support in Rapid, and adds support for Protomaps PMTiles URLs!

### Why

Overture Map Foundation has [recently released](https://overturemaps.org/overture-maps-foundation-releases-first-world-wide-open-map-dataset/) a whole lot of data under a permissive license.  Yay! 🎉   
Like everyone else, we're really interested to see how this data looks on a map. 
 
Shout out to @jenningsanderson and @bdon for building these demos that do exactly that:
- https://jenningsanderson.com/overture-daylight-mvp/#16.14/47.605721/-122.327055
- https://bdon.github.io/overture-tiles/places.html#16.14/47.605721/-122.327055

Both of these demos are powered by [Protomaps PMTiles](https://protomaps.com/docs/pmtiles) - a newish single-file format for storing vector tiles in the cloud.  It's like Mapbox Vector Tiles (MVT), but a map client can use HTTP Range Requests to get just the data you need from the archive. 

This PR adds support for .pmtiles sources in our "Custom Map Data" layers - this is the part of the Rapid code responsible for rendering custom .geojson, .kml, .gpx, files - and also connecting to vector tile sources of data.  

👉  **Importantly - the data is view-only for now.**  
This PR is not a solution for ingesting third party data into OSM, or bring-your-own-data, or anything like that.
You can connect to a PMTiles URL and look at what's in it, that's all!

Here are some .pmtiles URLs to try (no guarantees!):
```
https://jenningsanderson.com/overture-daylight-mvp/places.pmtiles
https://jenningsanderson.com/overture-daylight-mvp/buildings_z13.pmtiles
 https://jenningsanderson.com/overture-daylight-mvp/roads_z11.pmtiles
https://r2-public.protomaps.com/protomaps-sample-datasets/overture-pois.pmtiles
```

#### Overture Places around Newark, NJ
Screenshot 2023-08-10 at 4 20 51 PM

I also did some much-needed refresh of Rapid's vector tile code.  It's a lot more performant now, and it does a better job stitching together features across tile boundaries.  Take a peek into your favorite vector tile map!

#### OSM Americana around Newark, NJ
Screenshot 2023-08-10 at 4 29 42 PM

### What next
Many things that we can ticket separately
- bring-your-own-data #585
- improve/customize the styling of the vector data (it's all cyan for now)
- There are still bugs with stitching linestrings and polygons.

### CODE REVIEW

**Comment by bdon** [comment]
Nice! When will this make it to http://rapideditor.org ?

**Comment by bhousel** [comment]
> Nice! When will this make it to http://rapideditor.org ?

Sometime before end of August!  👍

**Comment by tordans** [comment]
FYI I tested the feature for an upcomming dataset which will be released officially soon which will include all parking spots in Berlin. We are considering how to migrate this data to OSM. Rapid could be a huge help for this.

Here is the sample dataset in Rapid https://rapideditor.org/edit#background=Berlin-2022&data=https://atlas-tiles.s3.eu-central-1.amazonaws.com/berlin-parking-polygons-euvm.pmtiles&datasets=fbRoads,msBuildings&disable_features=boundaries&map=19.76/52.49806/13.34018

(also a [preview of the pmtiles](https://protomaps.github.io/PMTiles/?url=https%3A%2F%2Fatlas-tiles.s3.eu-central-1.amazonaws.com%2Fberlin-parking-polygons-euvm.pmtiles#map=17.69/52.497235/13.332439) and a [preview along side our processed OSM parking data](https://radverkehrsatlas.de/regionen/parkraum?lat=52.4983&lng=13.3405&zoom=16.6&theme=parking&bg=default&config=!(i~parking~topics~!(i~parking~s~!(i~hidden~a~_F)(i~default~a)(i~presence~a~_F)(i~surface~a~_F))(i~parkingPoints~s~!(i~hidden~a)(i~default~a~_F))(i~parkingAreas~s~!(i~hidden~a~_F)(i~default~a)(i~street*_side~a~_F))(i~parkingDebug~s~!(i~hidden~a)(i~default~a~_F))(i~parkingStats~s~!(i~hidden~a)(i~stats-admin-level-4~a~_F)(i~default~a~_F)(i~stats-admin-level-10~a~_F)(i~length-admin-level-4~a~_F)(i~length-admin-level-9~a~_F)(i~length-admin-level-10~a~_F))(i~landuse~s~!(i~hidden~a)(i~default~a~_F)))~&data=!berlin-parking-polygons-euvm~))

The thing I am missing most right now in order to do anything with this feature in terms of mapping is a way to get the data in OSM. Something like a "create Geometry as OSM Polygon". And, of course, we would have to prepare the data to translate the existing tags to OSM tags in order to also have useful tags to keep (see also https://github.com/facebook/Rapid/issues/545).

**Comment by tordans** [comment]
> if someone's data includes a property like osm_tags or something similar, then we'd allow adding the feature to the OSM layer with those tags set. 

I like that idea. I think I would then keep the original attributes and only add the `osm_tags` property when I was able to create a confident set of tags. 

It would also make sure that not all data is added "blindly" but only such data that was crafted in the right way. Which I think is a good protection since one has to know a what to do in order to prepare the data and during that process there is enough time to think about licencing.

> We'd also probably need to remind them to make sure their data is compatible with the license. And maybe set an appropriate source tag.

I would appreciate some formality here. Maybe the "add to map" button is only present when an `osm_tags` but also a `osm_changeset_source` is given. And the source needs to be a osm wiki link which documents the data source, license and such. We used this approach when adding the parklets and other data I mentioned in https://github.com/facebook/Rapid/issues/585#issuecomment-1249994877 on the wiki pages linked unter "Qualitätssicherung der Daten" (EN: quality assurance of data)  https://wiki.openstreetmap.org/wiki/Berlin/Verkehrswende/Parkraum/Mapping_Kampagne_Xhain#Teil-Projekte_&_Aufgaben

---

What would the format of `osm_tags` be? Something simple like …?
```json
{ "properties": { "osm_keys": "key=value\nkey2=value2" } }
```

**Comment by tordans** [comment]
One more thing about the "add data to map" flow. A secondary use case would be to **update or merge data with an existing OSM object**. In the case of the parking datasets, we might have an existing parking polygon in OSM but want to update the tags with what we have in the external data. Maybe there is room for a feature where I select both the OSM and the external Data and instead of the "add shape with tags to map" button there is a "merge `osm_tags` with existing tags on selected osm object".

**Comment by RobJN** [comment]
I'm looking at PMTiles for OSMUK local chapter. I can see some cases where this would benefit us too, including the merge suggestion in the last comment.

**Comment by bdon** [comment]
One caveat I want to point out - for tiled vector sources in general, of which PMTiles overlays are one - is we need to be careful about how overzooming is handled. 

If the mapper is viewing Rapid at neighborhood scale and seeing z15 tiles, then the vector tile (MVT) geometry could be quite accurate. But if the vector tile datasource only contains zoom levels several before the viewed map (example: only tiles up to z10 or z11) then the geometry shown in Rapid may stray far from the intended positions.

This is equivalent to viewing satellite imagery as an overlay that has as limited zoom level and looks blurry when zoomed in. For raster tiles, the blur makes the problem obvious to the mapper, but I'm not sure how this would look in Rapid. If anyone is using tippecanoe to work with data in Rapid it would be helpful to show a demo of tilesets with fewer zoom levels.
